### PR TITLE
Add model for contact notes

### PR DESCRIPTION
**Pull request by ericnewcomer**

### CODE REVIEW

File: temba/api/v2/views.py [hunk 1]
```diff
@@ -1339,6 +1339,7 @@ def filter_queryset(self, queryset):
                 to_attr="prefetched_groups",
             ),
             Prefetch("current_flow"),
+            Prefetch("notes")
```
Comment (ericnewcomer) [inline temba/api/v2/views.py:1342]: @rowanseymour this doesn't seem right

Comment (rowanseymour) [reply temba/api/v2/views.py:1342]:
```python
Prefetch("notes", queryset=ContactNote.objects.order_by("id"))
```

File: temba/api/v2/serializers.py [hunk 2]
```diff
@@ -556,6 +557,15 @@ def get_groups(self, obj):
...
         return [{"uuid": g.uuid, "name": g.name} for g in groups]
 
+    def get_notes(self, obj):
+        if not obj.is_active:
+            return []
+        return [{
+            "text": note.text,
+            "created_on": note.created_on,
+            "created_by": note.created_by.username,
+        } for note in obj.notes.all().order_by("-pk")]
```
Comment (rowanseymour) [inline temba/api/v2/serializers.py:567]: get rid of order_by here

File: temba/contacts/models.py [hunk 3]
```diff
@@ -995,6 +995,16 @@ def apply_action_delete(cls, user, contacts):
...
             on_transaction_commit(lambda: release_contacts.delay(user.id, [c.id for c in contacts]))
 
+    def add_note(self, user, text):
+        """
+        Adds a note to this contact, prunes old ones if necessary
+        """
+        self.notes.create(text=text, created_by=user)
+
+        # remove all notes except the last 5
+        notes = self.notes.order_by("-pk").values_list("id", flat=True)[5:]
```
Comment (rowanseymour) [inline temba/contacts/models.py:1005]: we have a vague convention toward using `id` instead of `pk`

File: temba/api/v2/serializers.py [hunk 4]
```diff
@@ -594,6 +608,7 @@ class Meta:
 class ContactWriteSerializer(WriteSerializer):
     name = serializers.CharField(required=False, max_length=64, allow_null=True)
     language = serializers.CharField(required=False, min_length=3, max_length=3, allow_null=True)
+    note = serializers.CharField(required=False, allow_null=False)
```
Comment (rowanseymour) [inline temba/api/v2/serializers.py:611]: `allow_null=False` is redundant

Comment (rowanseymour) [reply temba/api/v2/serializers.py:611]: seems like we should have a `max_length` here which matches the UI... 10,000?

File: temba/contacts/models.py [hunk 5]
```diff
@@ -1678,6 +1688,13 @@ class Meta:
         constraints = [models.UniqueConstraint("org", Lower("name"), name="unique_contact_group_names")]
 
 
+class ContactNote(models.Model):
+    contact = models.ForeignKey(Contact, on_delete=models.PROTECT, related_name="notes")
+    text = models.TextField()
+    created_on = models.DateTimeField(auto_now_add=True)
```
Comment (rowanseymour) [inline temba/contacts/models.py:1694]: `models.DateTimeField(default=timezone.now)`

File: temba/contacts/tests.py [hunk 6]
```diff
@@ -1688,6 +1688,21 @@ def create_campaign(self):
...
         )
 
+    def test_contact_notes(self):
+        note_text = "This is note"
+
+        # create 10 notes
+        for i in range(10):
+            self.joe.add_note(self.user, f"{note_text} {i+1}")
+
+        notes = self.joe.notes.all().order_by("pk")
```
Comment (rowanseymour) [inline temba/contacts/tests.py:1698]: "id"

File: temba/api/v2/serializers.py [hunk 7]
```diff
@@ -556,6 +557,18 @@ def get_groups(self, obj):
...
         return [{"uuid": g.uuid, "name": g.name} for g in groups]
 
+    def get_notes(self, obj):
+        if not obj.is_active:
+            return []
+        return [
+            {
+                "text": note.text,
+                "created_on": note.created_on,
+                "created_by": note.created_by.username,
```
Comment (rowanseymour) [inline temba/api/v2/serializers.py:567]:
```python
"created_by": {"email": note.created_by.email, "name": note.created_by.name}
```

to match `UserField`

File: temba/api/v2/views.py [hunk 8]
```diff
@@ -1339,6 +1339,7 @@ def filter_queryset(self, queryset):
                 to_attr="prefetched_groups",
             ),
             Prefetch("current_flow"),
+            Prefetch("notes", queryset=ContactNote.objects.order_by("id")),
```
Comment (rowanseymour) [inline temba/api/v2/views.py:1342]: I think you also need

```python
Prefetch("notes__created_by")
```

File: temba/api/v2/tests.py [hunk 9]
```diff
@@ -1850,6 +1850,7 @@ def test_contacts(self, mr_mocks):
                 "language": "fra",
                 "urns": ["tel:+250788000004"],
                 "groups": [{"uuid": group.uuid, "name": group.name}],
+                "notes": [],
```
Comment (rowanseymour) [inline temba/api/v2/tests.py:1853]: should have at least one contact with a note to check prefetches

**Pull request merged**